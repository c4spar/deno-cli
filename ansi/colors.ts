import * as stdColors from "https://deno.land/std@0.82.0/fmt/colors.ts";

type ExcludedColorMethods = "setColorEnabled" | "getColorEnabled";
type PropertyNames = keyof typeof stdColors;
type ColorMethod = (str: string, ...args: Array<unknown>) => string;
type ColorMethods = Exclude<PropertyNames, ExcludedColorMethods>;
type Chainable<T, E extends keyof T | null = null> = {
  [P in keyof T]: P extends E ? T[P] : Chainable<T, E> & T[P];
};

/** Chainable colors instance returned by all ansi escape properties. */
export type ColorsChain = Chainable<typeof stdColors, ExcludedColorMethods>;

/** Create new `Colors` instance. */
export type ColorsFactory = () => Colors;

/**
 * Chainable colors module.
 * If invoked as method, a new `Colors` instance will be returned.
 */
export type Colors = ColorsFactory & ColorsChain;

/**
 * Chainable colors module.
 * ```
 * console.log(colors.blue.bgRed.bold('Welcome to Deno.Land!'));
 * ```
 * If invoked as method, a new Ansi instance will be returned.
 * ```
 * const myColors: Colors = colors();
 * console.log(myColors.blue.bgRed.bold('Welcome to Deno.Land!'));
 * ```
 */
export const colors: Colors = factory();

function factory(): Colors {
  let stack: Array<ColorMethods> = [];

  const colors: Colors = function (
    this: ColorsChain | undefined,
    str: string,
    ...args: Array<unknown>
  ): string | ColorsChain {
    if (this) {
      const lastname = stack.pop() as ColorMethods;
      const method = stdColors[lastname] as ColorMethod;
      const result: string = method(
        stack.reduce(
          (str: string, name) => (stdColors[name] as ColorMethod)(str),
          str,
        ),
        ...args,
      );
      stack = [];
      return result;
    }
    return factory();
  } as Colors;

  const methodNames = Object.keys(stdColors) as Array<PropertyNames>;

  for (const name of methodNames) {
    if (name === "setColorEnabled" || name === "getColorEnabled") {
      continue;
    }
    Object.defineProperty(colors, name, {
      get(this: ColorsChain) {
        stack.push(name);
        return this;
      },
    });
  }

  return colors;
}
