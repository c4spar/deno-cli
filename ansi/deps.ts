export { encodeBase64 } from "https://deno.land/std@0.216.0/encoding/base64.ts";
export type {
  ReaderSync,
  WriterSync,
} from "https://deno.land/std@0.216.0/io/types.ts";
