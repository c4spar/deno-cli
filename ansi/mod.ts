export * from "./ansi.ts";
export * from "./ansi_escapes.ts";
export * from "./tty.ts";
