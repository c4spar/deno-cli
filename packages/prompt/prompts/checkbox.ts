import { blue, dim, green, red } from 'https://deno.land/std@v0.41.0/fmt/colors.ts';
import { KeyEvent } from '../../keycode/mod.ts';
import { Figures } from '../lib/figures.ts';
import { GenericList, GenericListItemOptions, GenericListItemSettings, GenericListNamedItemOptions, GenericListPromptOptions, GenericListPromptSettings } from './generic-list.ts';

export interface CheckboxItemOptions extends GenericListItemOptions {
    checked?: boolean;
    icon?: boolean;
}

export interface CheckboxNamedItemOptions extends GenericListNamedItemOptions {
    checked?: boolean;
    icon?: boolean;
}

export interface CheckboxItemSettings extends GenericListItemSettings {
    checked: boolean;
    icon: boolean;
}

export type CheckboxValueOptions =
    ( string | CheckboxNamedItemOptions )[]
    | { [ s: string ]: string | CheckboxItemOptions };
export type CheckboxValueSettings = CheckboxItemSettings[];

export interface CheckboxPromptOptions extends GenericListPromptOptions<string[], string[]> {
    indent?: string;
    pointer?: string;
    check?: string;
    uncheck?: string;
    values: CheckboxValueOptions;
}

export interface CheckboxPromptSettings extends GenericListPromptSettings<string[], string[]> {
    indent: string;
    pointer: string;
    check: string;
    uncheck: string;
    values: CheckboxValueSettings;
}

export class Checkbox extends GenericList<string[], string[], CheckboxPromptSettings> {

    public static async prompt( options: CheckboxPromptOptions ): Promise<string[] | undefined> {

        const items: CheckboxNamedItemOptions[] = this.mapValues( options.values );
        const values: CheckboxValueSettings = items.map( item => this.mapItem( item, options.default ) );

        return new this( {
            pointer: blue( Figures.POINTER ),
            indent: ' ',
            maxRows: 0,
            check: green( Figures.TICK ),
            uncheck: red( Figures.CROSS ),
            ...options,
            values
        } ).run();
    }

    public static separator( label: string = '------------' ): CheckboxNamedItemOptions {
        return {
            ...super.separator(),
            icon: false
        };
    }

    protected static mapValues( optValues: CheckboxValueOptions ): CheckboxNamedItemOptions[] {
        return super.mapValues( optValues ) as CheckboxNamedItemOptions[];
    }

    protected static mapItem( item: CheckboxNamedItemOptions, defaults?: string[] ): CheckboxItemSettings {
        return {
            ...super.mapItem( item ),
            checked: typeof item.checked === 'undefined' && defaults && defaults.indexOf( item.name ) !== -1 ? true : !!item.checked,
            icon: typeof item.icon === 'undefined' ? true : item.icon
        };
    }

    protected getMessage(): string {

        let message = this.settings.message;

        if ( typeof this.settings.default !== 'undefined' ) {
            message += dim( ` (${ this.settings.default.join( ', ' ) })` );
        }

        return message;
    }

    protected async handleEvent( event: KeyEvent ): Promise<boolean> {

        switch ( event.name ) {

            case 'c':
                if ( event.ctrl ) {
                    return Deno.exit( 0 );
                }
                break;

            case 'up':
                await this.selectPrevious();
                break;

            case 'down':
                await this.selectNext();
                break;

            case 'space':
                this.checkValue();
                break;

            case 'return':
            case 'enter':
                return this.selectValue();
        }

        return false;
    }

    protected checkValue() {
        const item = this.settings.values[ this.selected ];
        item.checked = !item.checked;
    }

    protected values() {
        return this.settings.values
                   .filter( item => item.checked )
                   .map( item => item.name );
    }

    protected async selectValue() {
        return this.validateValue( this.values() );
    }

    protected writeListItem( item: CheckboxItemSettings, isSelected?: boolean ) {

        let line = this.settings.indent;

        // pointer
        line += isSelected ? `${ this.settings.pointer } ` : '  ';

        // icon
        if ( item.icon ) {
            let check = item.checked ? `${ this.settings.check } ` : `${ this.settings.uncheck } `;
            if ( item.disabled ) {
                check = dim( check );
            }
            line += check;
        } else {
            line += '  ';
        }

        // value
        const value: string = item.label;
        line += `${ isSelected ? value : dim( value ) }`;

        this.writeLine( line );
    }

    protected sanitize( value: string[] ): string[] {
        return value;
    }

    protected validate( value: string[] ): boolean {
        return true;
    }

    protected transform( value: string[] ): string {
        return value.map( val => val.slice( 0, 1 ).toUpperCase() + val.slice( 1 ) )
                    .join( ' ' );
    }
}
