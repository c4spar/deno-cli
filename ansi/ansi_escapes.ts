import { encodeBase64 } from "./deps.ts";

/** Image options. */
export interface ImageOptions {
  width?: number;
  height?: number;
  preserveAspectRatio?: boolean;
}

/** Escape sequence: `\x1B` */
const ESC = "\x1B";
/** Control sequence intro: `\x1B[` */
const CSI = `${ESC}[`;
/** Operating system command: `\x1B]` */
const OSC = `${ESC}]`;
/** Link separator */
const SEP = ";";

export const ansiEscapes = {
  /** Ring audio bell: `\u0007` */
  bel: "\u0007",
  /** Get cursor position. */
  cursorPosition: `${CSI}6n`,
  /**
   * Move cursor to x, y, counting from the top left corner.
   * @param x Position left.
   * @param y Position top.
   */
  cursorTo(x: number, y?: number): string {
    if (typeof y !== "number") {
      return `${CSI}${x}G`;
    }
    return `${CSI}${y};${x}H`;
  },
  /**
   * Move cursor by offset.
   * @param x Offset left.
   * @param y Offset top.
   */
  cursorMove(x: number, y: number): string {
    let ret = "";

    if (x < 0) {
      ret += `${CSI}${-x}D`;
    } else if (x > 0) {
      ret += `${CSI}${x}C`;
    }

    if (y < 0) {
      ret += `${CSI}${-y}A`;
    } else if (y > 0) {
      ret += `${CSI}${y}B`;
    }

    return ret;
  },
  /**
   * Move cursor up by n lines.
   * @param count Number of lines.
   */
  cursorUp: (count = 1): string => `${CSI}${count}A`,
  /**
   * Move cursor down by n lines.
   * @param count Number of lines.
   */
  cursorDown: (count = 1): string => `${CSI}${count}B`,
  /**
   * Move cursor forward by n lines.
   * @param count Number of lines.
   */
  cursorForward: (count = 1): string => `${CSI}${count}C`,
  /**
   * Move cursor backward by n lines.
   * @param count Number of lines.
   */
  cursorBackward: (count = 1): string => `${CSI}${count}D`,
  /**
   * Move cursor to the beginning of the line n lines down.
   * @param count Number of lines.
   */
  cursorNextLine: (count = 1): string => `${CSI}E`.repeat(count),
  /**
   * Move cursor to the beginning of the line n lines up.
   * @param count Number of lines.
   */
  cursorPrevLine: (count = 1): string => `${CSI}F`.repeat(count),
  /** Move cursor to first column of current row. */
  cursorLeft: `${CSI}G`,
  /** Hide cursor. */
  cursorHide: `${CSI}?25l`,
  /** Show cursor. */
  cursorShow: `${CSI}?25h`,
  /** Save cursor. */
  cursorSave: `${ESC}7`,
  /** Restore cursor. */
  cursorRestore: `${ESC}8`,
  /**
   * Scroll window up by n lines.
   * @param count Number of lines.
   */
  scrollUp: (count = 1): string => `${CSI}S`.repeat(count),
  /**
   * Scroll window down by n lines.
   * @param count Number of lines.
   */
  scrollDown: (count = 1): string => `${CSI}T`.repeat(count),
  /** Clear screen. */
  eraseScreen: `${CSI}2J`,
  /**
   * Clear screen up.
   * @param count Number of lines.
   */
  eraseUp: (count = 1): string => `${CSI}1J`.repeat(count),
  /**
   * Clear screen down.
   * @param count Number of lines.
   */
  eraseDown: (count = 1): string => `${CSI}0J`.repeat(count),
  /** Clear current line. */
  eraseLine: `${CSI}2K`,
  /** Clear to line end. */
  eraseLineEnd: `${CSI}0K`,
  /** Clear to line start. */
  eraseLineStart: `${CSI}1K`,
  /**
   * Clear screen and move cursor by n lines up and move cursor to first column.
   * @param count Number of lines.
   */
  eraseLines(count: number): string {
    let clear = "";
    for (let i = 0; i < count; i++) {
      clear += ansiEscapes.eraseLine +
        (i < count - 1 ? ansiEscapes.cursorUp() : "");
    }
    clear += ansiEscapes.cursorLeft;
    return clear;
  },
  /**
   * Create link.
   * @param text Link text.
   * @param url Link url.
   * ```
   * console.log(
   *   ansi.link("Click me.", "https://deno.land"),
   * );
   * ```
   */
  link: (text: string, url: string): string =>
    [
      OSC,
      "8",
      SEP,
      SEP,
      url,
      ansiEscapes.bel,
      text,
      OSC,
      "8",
      SEP,
      SEP,
      ansiEscapes.bel,
    ].join(""),
  /**
   * Create image.
   * @param buffer  Image buffer.
   * @param options Image options.
   * ```
   * const response = await fetch("https://deno.land/images/hashrock_simple.png");
   * const imageBuffer: ArrayBuffer = await response.arrayBuffer();
   * console.log(
   *   ansi.image(imageBuffer),
   * );
   * ```
   */
  image(
    buffer: string | ArrayBuffer,
    options?: ImageOptions,
  ): string {
    let ret = `${OSC}1337;File=inline=1`;

    if (options?.width) {
      ret += `;width=${options.width}`;
    }

    if (options?.height) {
      ret += `;height=${options.height}`;
    }

    if (options?.preserveAspectRatio === false) {
      ret += ";preserveAspectRatio=0";
    }

    return ret + ":" + encodeBase64(buffer) + ansiEscapes.bel;
  },
};
