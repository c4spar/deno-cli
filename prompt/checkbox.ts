import type { KeyCode } from "../keycode/mod.ts";
import { brightBlue, dim, green, red } from "./deps.ts";
import { Figures, getFiguresByKeys } from "./figures.ts";
import {
  GenericList,
  GenericListKeys,
  GenericListOption,
  GenericListOptionGroup,
  GenericListOptionGroupSettings,
  GenericListOptions,
  GenericListOptionSettings,
  GenericListSettings,
  isOption,
  isOptionGroup,
} from "./_generic_list.ts";
import { GenericPrompt } from "./_generic_prompt.ts";

/** Checkbox prompt options. */
export interface CheckboxOptions
  extends GenericListOptions<Array<string>, Array<string>> {
  options: Array<string | CheckboxOption | CheckboxOptionGroup>;
  confirmSubmit?: boolean;
  check?: string;
  uncheck?: string;
  partialCheck?: string;
  minOptions?: number;
  maxOptions?: number;
  keys?: CheckboxKeys;
}

/** Checkbox prompt settings. */
interface CheckboxSettings extends
  GenericListSettings<
    Array<string>,
    Array<string>,
    CheckboxOptionSettings,
    CheckboxOptionGroupSettings
  > {
  confirmSubmit: boolean;
  check: string;
  uncheck: string;
  partialCheck: string;
  minOptions: number;
  maxOptions: number;
  keys: CheckboxKeys;
}

/** Checkbox option options. */
export interface CheckboxOption extends GenericListOption {
  checked?: boolean;
  icon?: boolean;
}

/** Checkbox option group options. */
export interface CheckboxOptionGroup
  extends GenericListOptionGroup<CheckboxOption> {
  icon?: boolean;
}

/** Checkbox option settings. */
export interface CheckboxOptionSettings extends GenericListOptionSettings {
  checked: boolean;
  icon: boolean;
}

/** Checkbox option group settings. */
export interface CheckboxOptionGroupSettings
  extends GenericListOptionGroupSettings<CheckboxOptionSettings> {
  readonly checked: boolean;
  icon: boolean;
}

/** Checkbox key options. */
export interface CheckboxKeys extends GenericListKeys {
  check?: Array<string>;
}

/** Checkbox prompt representation. */
export class Checkbox extends GenericList<
  Array<string>,
  Array<string>,
  CheckboxOptionSettings,
  CheckboxOptionGroupSettings
> {
  protected readonly settings: CheckboxSettings;
  protected options: Array<
    CheckboxOptionSettings | CheckboxOptionGroupSettings
  >;
  protected listIndex: number;
  protected listOffset: number;
  private confirmSubmit = false;

  /** Execute the prompt and show cursor on end. */
  public static prompt(options: CheckboxOptions): Promise<Array<string>> {
    return new this(options).prompt();
  }

  /**
   * Create list separator.
   * @param label Separator label.
   */
  public static separator(label?: string): CheckboxOption {
    return {
      ...super.separator(label),
      icon: false,
    };
  }

  /**
   * Inject prompt value. Can be used for unit tests or pre selections.
   * @param value Array of input values.
   */
  public static inject(value: Array<string>): void {
    GenericPrompt.inject(value);
  }

  constructor(options: CheckboxOptions) {
    super();
    this.settings = this.getDefaultSettings(options);
    this.options = this.settings.options.slice();
    this.listIndex = this.getListIndex();
    this.listOffset = this.getPageOffset(this.listIndex);
  }

  protected getDefaultSettings(options: CheckboxOptions): CheckboxSettings {
    const settings = super.getDefaultSettings(options);
    return {
      confirmSubmit: true,
      ...settings,
      check: options.check ?? green(Figures.TICK),
      uncheck: options.uncheck ?? red(Figures.CROSS),
      partialCheck: options.partialCheck ?? green(Figures.RADIO_ON),
      minOptions: options.minOptions ?? 0,
      maxOptions: options.maxOptions ?? Infinity,
      options: this.mapOptions(options, options.options),
      keys: {
        check: ["space"],
        ...(settings.keys ?? {}),
        open: options.keys?.open ?? ["right"],
        back: options.keys?.back ?? ["left", "escape"],
      },
    };
  }

  /**
   * Map string option values to options and set option defaults.
   * @param promptOptions Checkbox options.
   */
  protected mapOptions(
    promptOptions: CheckboxOptions,
    options: Array<string | CheckboxOption | CheckboxOptionGroup>,
  ): Array<CheckboxOptionSettings | CheckboxOptionGroupSettings> {
    return options.map((option) =>
      isOptionGroup(option)
        ? this.mapOptionGroup(promptOptions, option)
        : typeof option === "string"
        ? this.mapOption(promptOptions, { value: option })
        : this.mapOption(promptOptions, option)
    );
  }

  /**
   * Set list option defaults.
   * @param option List option.
   */
  protected mapOption(
    options: CheckboxOptions,
    option: CheckboxOption,
  ): CheckboxOptionSettings {
    return {
      ...super.mapOption(options, option),
      checked: typeof option.checked === "undefined" && options.default &&
          options.default.indexOf(option.value) !== -1
        ? true
        : !!option.checked,
      icon: typeof option.icon === "undefined" ? true : option.icon,
    };
  }

  protected mapOptionGroup(
    promptOptions: CheckboxOptions,
    option: CheckboxOptionGroup,
  ): CheckboxOptionGroupSettings {
    const options = this.mapOptions(promptOptions, option.options);
    return {
      ...super.mapOptionGroup(promptOptions, option, false),
      get checked() {
        return areAllChecked(options);
      },
      options,
      icon: typeof option.icon === "undefined" ? true : option.icon,
    };
  }

  protected match(): void {
    super.match();
    if (this.isSearching()) {
      this.selectSearch();
    }
  }

  protected getListItemIcon(
    option: CheckboxOptionSettings | CheckboxOptionGroupSettings,
  ): string {
    return this.getCheckboxIcon(option) + super.getListItemIcon(option);
  }

  private getCheckboxIcon(
    option: CheckboxOptionSettings | CheckboxOptionGroupSettings,
  ): string {
    if (!option.icon) {
      return "";
    }
    const icon = option.checked
      ? this.settings.check + " "
      : isOptionGroup(option) && areSomeChecked(option.options)
      ? this.settings.partialCheck + " "
      : this.settings.uncheck + " ";

    return option.disabled ? dim(icon) : icon;
  }

  /** Get value of checked options. */
  protected getValue(): Array<string> {
    return flatOptions<CheckboxOptionSettings, CheckboxOptionGroupSettings>(
      this.settings.options,
    )
      .filter((option) => option.checked)
      .map((option) => option.value);
  }

  /**
   * Handle user input event.
   * @param event Key event.
   */
  protected async handleEvent(event: KeyCode): Promise<void> {
    const hasConfirmed: boolean = this.confirmSubmit;
    this.confirmSubmit = false;

    switch (true) {
      case this.isKey(this.settings.keys, "check", event) &&
        !this.isSearchSelected():
        this.checkValue();
        break;
      case this.isKey(this.settings.keys, "submit", event):
        await this.submit(hasConfirmed);
        break;
      default:
        await super.handleEvent(event);
    }
  }

  protected hint(): string | undefined {
    if (this.confirmSubmit) {
      const info = this.selectedOption && this.isBackButton(this.selectedOption)
        ? ` To leave the group use ${
          getFiguresByKeys(this.settings.keys.back ?? []).join(", ")
        }.`
        : ` To open the selected group ${
          getFiguresByKeys(this.settings.keys.open ?? []).join(", ")
        }.`;

      return this.settings.indent +
        brightBlue(
          `Press ${
            getFiguresByKeys(this.settings.keys.submit ?? [])
          } again to submit.${info}`,
        );
    }

    return super.hint();
  }

  protected async submit(hasConfirmed?: boolean): Promise<void> {
    if (
      !hasConfirmed &&
      this.settings.confirmSubmit &&
      Deno.isatty(Deno.stdout.rid) &&
      !this.isSearchSelected()
    ) {
      this.confirmSubmit = true;
      return;
    }

    await super.submit();
  }

  /** Check selected option. */
  protected checkValue(): void {
    const option = this.options.at(this.listIndex);
    if (!option) {
      this.setErrorMessage("No option available to select.");
      return;
    } else if (option.disabled) {
      this.setErrorMessage("This option is disabled and cannot be changed.");
      return;
    }
    this.checkOption(option, !option.checked);
  }

  private checkOption(
    option: CheckboxOptionSettings | CheckboxOptionGroupSettings,
    checked: boolean,
  ) {
    if (isOption(option)) {
      option.checked = checked;
    } else {
      for (const childOption of option.options) {
        this.checkOption(childOption, checked);
      }
    }
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: Array<string>): boolean | string {
    const options = flatOptions<
      CheckboxOptionSettings,
      CheckboxOptionGroupSettings
    >(this.settings.options);
    const isValidValue = Array.isArray(value) &&
      value.every((val) =>
        typeof val === "string" &&
        val.length > 0 &&
        options.findIndex((option: CheckboxOptionSettings) =>
            option.value === val
          ) !== -1
      );

    if (!isValidValue) {
      return false;
    }

    if (value.length < this.settings.minOptions) {
      return `The minimum number of options is ${this.settings.minOptions} but got ${value.length}.`;
    }
    if (value.length > this.settings.maxOptions) {
      return `The maximum number of options is ${this.settings.maxOptions} but got ${value.length}.`;
    }

    return true;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: Array<string>): Array<string> {
    return value.map((val) => val.trim());
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: Array<string>): string {
    return value.map((val) => this.getOptionByValue(val)?.name ?? val)
      .join(", ");
  }
}

function areSomeChecked(
  options: Array<CheckboxOptionSettings | CheckboxOptionGroupSettings>,
): boolean {
  return options.some((option) =>
    isOptionGroup(option) ? areSomeChecked(option.options) : option.checked
  );
}

function areAllChecked(
  options: Array<CheckboxOptionSettings | CheckboxOptionGroupSettings>,
): boolean {
  return options.every((option) =>
    isOptionGroup(option) ? areAllChecked(option.options) : option.checked
  );
}

function flatOptions<
  TOption extends GenericListOptionSettings,
  TGroup extends GenericListOptionGroupSettings<TOption>,
>(
  options: Array<TOption | TGroup>,
): Array<TOption> {
  return flat(options);

  function flat(
    options: Array<TOption | TGroup>,
    indentLevel = 0,
    opts: Array<TOption> = [],
  ): Array<TOption> {
    for (const option of options) {
      option.indentLevel = indentLevel;
      if (isOption(option)) {
        opts.push(option);
      }
      if (isOptionGroup(option)) {
        flat(option.options, ++indentLevel, opts);
      }
    }

    return opts;
  }
}
/**
 * Checkbox options type.
 * @deprecated Use `Array<string | CheckboxOption>` instead.
 */
export type CheckboxValueOptions = Array<string | CheckboxOption>;

/**
 * Checkbox option settings type.
 * @deprecated Use `Array<CheckboxOptionSettings | CheckboxOptionGroupSettings>` instead.
 */
export type CheckboxValueSettings = Array<
  CheckboxOptionSettings | CheckboxOptionGroupSettings
>;
