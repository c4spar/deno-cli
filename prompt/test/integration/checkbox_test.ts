import { ansi } from "../../../ansi/ansi.ts";
import { Checkbox } from "../../checkbox.ts";
import { snapshotTest } from "../../../testing/snapshot.ts";

await snapshotTest({
  name: "checkbox prompt > should check an option",
  meta: import.meta,
  osSuffix: ["windows"],
  stdin: ansi
    .cursorDown
    .cursorDown
    .text(" ")
    .text("\n")
    .toArray(),
  async fn() {
    await Checkbox.prompt({
      message: "Select an option",
      options: [
        { name: "Foo", value: "foo" },
        { name: "Bar", value: "bar" },
        { name: "Baz", value: "baz" },
      ],
    });
  },
});

await snapshotTest({
  name: "checkbox prompt > should search an option",
  meta: import.meta,
  osSuffix: ["windows"],
  stdin: ansi
    .text("baz")
    .text("\n")
    .text(" ")
    .text("\n")
    .toArray(),
  async fn() {
    await Checkbox.prompt({
      message: "Select an option",
      search: true,
      options: [
        { name: "Foo", value: "foo" },
        { name: "Bar", value: "bar" },
        { name: "Baz", value: "baz" },
      ],
    });
  },
});
