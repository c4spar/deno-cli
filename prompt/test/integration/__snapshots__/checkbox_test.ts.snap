export const snapshot = {};

snapshot[`checkbox prompt > should check an option 1`] = `
stdout:
"? Select an option
❯ ✘ Foo
  ✘ Bar
  ✘ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
❯ ✘ Bar
  ✘ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
  ✘ Bar
❯ ✘ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  ✘ Foo
  ✘ Bar
❯ ✔ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option › Baz
\\x1b[?25h\\x1b[?25h"
stderr:
""
`;

snapshot[`checkbox prompt > should search an option 1`] = `
stdout:
"? Select an option 🔎 
❯ ✘ Foo
  ✘ Bar
  ✘ Baz\\x1b[3A\\x1b[23G\\x1b[G\\x1b[0J
? Select an option 🔎 baz
  ✘ Baz\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? Select an option 🔎 baz
❯ ✘ Baz\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? Select an option 🔎 baz
❯ ✔ Baz\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? Select an option › Baz
\\x1b[?25h\\x1b[?25h"
stderr:
""
`;
